www/waterfall_view/src/module/main.module.js

```diff
@@ -635,14 +640,89 @@ var WaterfallController = (function() {
             });
         }
 
-        renderNewData() {
+        toggleTag(tag) {
+            if (!this.$scope.tags_filter.includes(tag)) {
+                this.$scope.tags_filter.push(tag);
+            } else {
+                this.$scope.tags_filter = _.remove(this.$scope.tags_filter, function(currentTag) {
+                    return currentTag != tag;
+                });
+            }
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        isTagToggled(tag) {
+            if (this.$scope.tags_filter.includes(tag)) {
+                return 'label-success';
+            } else {
+                return 'label-default';
+            }
+        }
+
+        clearTags() {
+            this.$scope.tags_filter = this.tags_filter = [];
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        renderNewData(currentTags) {
+            currentTags = this.$location.search()['tags'];
+            if (currentTags != null) {
+                if (typeof currentTags == 'string'){
+                    this.$scope.tags_filter = this.tags_filter = [currentTags];
```

Comment:
We should set bidirectional data biding shouldn't we? @tardyp I'm not 100% confident in my understanding of AngularJS concepts.

Reply:
I tried using `$rootScope.$on('$locationChangeSuccess', ()=>{})` but inside that I wasn't able to change `$scope.tags_filter`, so I put the logic inside `renderNewData`

Reply:
I think you should test more on locationChangeSuccess.

For example, if you bookmark and jump directly, it does not work on the test server:
http://ec2-18-191-33-221.us-east-2.compute.amazonaws.com/#/waterfall?tags=nodejs10

here there is not filtering whereas we should expect to see only nodejs10 builds